testutil/testpolylog/event.go

```diff
@@ -0,0 +1,166 @@
+package testpolylog
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	"github.com/pokt-network/poktroll/pkg/polylog"
+)
+
+type (
+	// NewLoggerAndOutputFn is called in the test helper to create a new logger
+	// configured with the given level and options. It returns the logger and
+	// the buffer to which the logger writes. It is useful for decoupling test
+	// helpers from a specific logger implementation.
+	NewLoggerAndOutputFn func(
+		*testing.T,
+		polylog.Level,
+		...polylog.LoggerOption,
+	) (polylog.Logger, *bytes.Buffer)
+
+	// NewEventWithLevelFn is called in the test helper to create a new event
+	// at the given level from the given logger. It is useful for decoupling
+	// test helpers from a specific logger implementation.
+	NewEventWithLevelFn func(
+		*testing.T,
+		polylog.Logger,
+		polylog.Level,
+	) polylog.Event
+)
+
+// FnMethodSpy is a mock which implements a #Fn() method that is intended to be
+// used in tests to assert that the function passed to polylog.Event#Func() is
+// called with the expected arg(s).
+type FnMethodSpy struct{ mock.Mock }
+
+// Fn is a mock method which can be asserted on via the mock.Mock API.
+// See: https://pkg.go.dev/github.com/stretchr/testify@v1.8.4/mock#Mock.
+func (m *FnMethodSpy) Fn(event polylog.Event) {
+	m.Called(event)
+}
+
+// EventMethodsTest is a test case for expressing and exercising polylog.Event
+// methods in a concise way.
+type EventMethodsTest struct {
+	Msg                    string
+	MsgFmt                 string
+	MsgFmtArgs             []any
+	Key                    string
+	Value                  any
+	EventMethodName        string
+	ExpectedOutputContains string
+}
+
+// RunEventMethodTests runs a set of tests for a given level.
+func RunEventMethodTests(
+	t *testing.T,
+	level polylog.Level,
+	tests []EventMethodsTest,
+	newLoggerAndOutput NewLoggerAndOutputFn,
+	newEventWithLevel NewEventWithLevelFn,
+	funcMethodEventTypeName string,
+	getExpectedLevelOutputContains func(level polylog.Level) string,
+) {
+	t.Helper()
+
+	// Title-case level string so that it can be used as the name of the
+	// method to call on the logger using reflect and for the sub-test
+	// descriptions.
+	//
+	// TODO_TECHDEBT/TODO_COMMUNITY: `strings.Title()` is deprecated. Follow
+	// migration guidance in godocs: https://pkg.go.dev/strings@go1.21.4#Title.
+	levelMethodStr := strings.Title(level.String())
+
+	for _, tt := range tests {
+		// If the test case does not specify an event method name, use the test
+		// case's key as the event method name. This is done for convenience only.
```

Comment:
`This is done for convenience only.`

Is it hard to avoid? I want to try to avoid easy things this early in the design of a new testing primitive.

Reply:
I'll push back on this. I think the added verbosity needing to explicitly specify this will remove potential confusion/indirection from test readers/authors.